### PR TITLE
Webpack watch mode.

### PR DESCRIPTION
**Pull request by nicolad**
##  Problem this Pull Request solves
- Adds watch script.

### CODE REVIEW

File: config/webpack.config.js [hunk 1]
```diff
@@ -616,13 +617,6 @@ module.exports = function(webpackEnv) {
 					// The formatter is invoked directly in WebpackDevServerUtils during development
 					formatter: isEnvProduction ? typescriptFormatter : undefined,
 				}),
-			isEnvDevelopment &&
-				new WriteFilePlugin({
-					// Write only files that have ".ts" extension.
-					test: /\.ts$/,
-					useHashIndex: true,
-				}),
-
```
Comment (nicolad) [inline config/webpack.config.js]: That wasn't helping in any way.

File: config/webpack.config.js [hunk 2]
```diff
@@ -144,9 +143,7 @@ module.exports = function(webpackEnv) {
 	};
 
 	const entry = Object.entries(entries).reduce((newObj, [entry, path]) => {
-		newObj[entry] = [isEnvDevelopment && require.resolve('react-dev-utils/webpackHotDevClient'), path].filter(
-			Boolean
-		);
```
Comment (nicolad) [inline config/webpack.config.js]: Since we are not using HMR, we don't need `'react-dev-utils/webpackHotDevClient'`

File: config/webpack.config.js [hunk 3]
```diff
@@ -171,7 +168,7 @@ module.exports = function(webpackEnv) {
 
 		output: {
 			// The build folder.
-			path: isEnvProduction ? paths.appBuild : undefined,
+			path: paths.appBuild,
```
Comment (nicolad) [inline config/webpack.config.js]: In dev it was outputting in `./dist` instead of `./assets/dist`; this is why this change has been made.

File: package.json [hunk 4]
```diff
@@ -29,12 +29,12 @@
...
 		"lint-php:skip-warnings": "composer run-script lint:skip-warnings",
 		"pretty": "prettier --write \"assets/prototype/**/*.js\"",
-		"prewatch": "npm run check-engines",
 		"test": "npm run lint && npm run test-js",
 		"test-php": "cd tests && phpunit",
 		"test-js": "node scripts/test.js",
 		"test-js:coverage": "npm run test-js -- --coverage",
-		"test-js:coverage-ci": "npm run test-js -- --coverage --maxWorkers 1"
+		"test-js:coverage-ci": "npm run test-js -- --coverage --maxWorkers 1",
+		"watch": "node scripts/watch.js"
```
Comment (nicolad) [inline package.json]: We might not need `dev` command or we might replace that one with this, not sure.

File: scripts/watch.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+		})
+	);
+});
+
+function copyPublicFolder() {
+	fs.copySync(paths.appPublic, paths.appBuild, {
+		dereference: true,
+		filter: (file) => file !== paths.appHtml,
+	});
+}
```
Comment (nicolad) [inline scripts/watch.js]: This is inspired from: https://gist.github.com/int128/e0cdec598c5b3db728ff35758abdbafd

File: config/webpack.config.js [hunk 6]
```diff
@@ -171,7 +168,7 @@ module.exports = function(webpackEnv) {
 
 		output: {
 			// The build folder.
-			path: isEnvProduction ? paths.appBuild : undefined,
```
Comment (manzoorwanijk) [inline config/webpack.config.js]: Nice hack 😄

File: scripts/watch.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+const paths = require('../config/paths');
+const webpack = require('webpack');
+const config = require('../config/webpack.config.js');
+const devConfig = config(DEVELOPMENT);
+
+webpack(devConfig).watch({}, (err, stats) => {
+	if (err) {
+		console.error(err);
+	}
+	console.error(
```
Comment (manzoorwanijk) [inline scripts/watch.js]: Is it supposed to be `console.error()`?

Comment (nicolad) [reply scripts/watch.js]: Not sure, maybe they wanted to emphasize.
Updated to be `console.log`
